src/pd/client.rs

```diff
@@ -11,247 +11,405 @@
+            Ok(c) => c,
+            // Ignore failed PD node.
+            Err(e) => {
+                error!("PD endpoint {} is down: {:?}", ep, e);
+                continue;
+            }
+        };
 
-                Err(_) => {
-                    error!("failed to connect to {}, try next", ep);
-                    continue;
-                }
+        let resp = match client.GetMembers(pdpb::GetMembersRequest::new()) {
+            Ok(resp) => resp,
+            // Ignore failed PD node.
+            Err(e) => {
+                error!("PD endpoint {} failed to respond: {:?}", ep, e);
+                continue;
             }
+        };
+
+        // Check cluster ID.
+        let cid = resp.get_header().get_cluster_id();
+        if let Some(sample) = cluster_id {
+            if sample != cid {
+                return Err(box_err!("PD response cluster_id mismatch, want {}, got {}",
+                                    sample,
+                                    cid));
+            }
+        } else {
+            cluster_id = Some(cid);
         }
+        // TODO: check all fields later?
 
-        Err(box_err!("failed to connect to {:?}", self.endpoints))
+        if members.is_none() {
+            members = Some(resp);
+        }
     }
 
-    fn send(&mut self, msg_id: u64, req: &Request) -> Result<Response> {
-        // If we post failed, we should retry.
-        for _ in 0..MAX_PD_SEND_RETRY_COUNT {
-            // If no stream, try connect first.
-            if self.stream.is_none() && self.try_connect().is_err() {
-                // TODO: figure out a better way to do backoff
-                thread::sleep(Duration::from_millis(50));
-                continue;
-            }
+    match members {
+        Some(members) => {
+            let (client, members) = try!(try_connect_leader(&members));
+            info!("All PD endpoints are consistent: {:?}", endpoints);
+            Ok((client, members))
+        }
+        _ => Err(box_err!("PD cluster failed to respond")),
+    }
+}
 
-            let mut stream = self.stream.take().unwrap();
-            // We may send message to a not leader pd, retry.
+fn connect(addr: &str) -> Result<PDClient> {
+    debug!("connect to PD endpoint: {:?}", addr);
+    let ep = box_try!(Url::parse(addr));
+    let host = match ep.host_str() {
+        Some(h) => h.to_owned(),
+        None => return Err(box_err!("unkown host, please specify the host")),
+    };
+    let port = match ep.port() {
+        Some(p) => p,
+        None => return Err(box_err!("unkown port, please specify the port")),
+    };
+
+    let mut conf: grpc::client::GrpcClientConf = Default::default();
+    conf.http.no_delay = Some(true);
+
+    // TODO: It seems that `new` always return an Ok(_).
+    PDClient::new(&host, port, false, conf)
+        .and_then(|client| {
+            // try request.
+            match client.GetMembers(pdpb::GetMembersRequest::new()) {
+                Ok(_) => Ok(client),
```

Comment:
This function only creates a client, `new` always returns a Ok(_), `GetMembers` here is for testing whether the client is OK or not. it's a workaround. See more at https://github.com/pingcap/tikv/pull/1591#discussion_r106353944